[Excel]  (Custom functions) Combine localization and naming articles (#4451)
OfficeDev/office-js-docs-pr: docs/excel/custom-functions-naming.md

[!include[manifest guidance](../includes/manifest-guidance.md)]

## Localize custom functions

You can localize both your add-in and your custom function names. To do so, provide localized function names in the functions' JSON file and locale information in the XML manifest file.

> >[!IMPORTANT]

Reviewer: ```suggestion
> [!IMPORTANT]
```

> > Autogenerated metadata doesn't work for localization so you need to update the JSON file manually. To learn how to do this, see [Manually create JSON metadata for custom functions](custom-functions-json.md)

Reviewer: ```suggestion
> Autogenerated metadata doesn't work for localization so you need to update the JSON file manually. To learn how to do this, see [Manually create JSON metadata for custom functions](custom-functions-json.md).
```

### Localize function names

> To localize your custom functions, create a new JSON metadata file for each language. In each language JSON file, create `name` and `description` properties in the target language. The default file for English is named **functions.json**. Use the locale in the filename for each additional JSON file, such as **functions-de.json** to help identify them.

Reviewer: Suggestion to avoid repeating "create".
```suggestion
To localize your custom functions, create a new JSON metadata file for each language. In each language JSON file, add `name` and `description` properties in the target language. The default file for English is named **functions.json**. Use the locale in the filename for each additional JSON file, such as **functions-de.json** to help identify them.
```

The `name` and `description` appear in Excel and are localized. However, the `id` of each function isn't localized. The `id` property is how Excel identifies your function as unique and shouldn't be changed once it is set.

> >[!IMPORTANT]

Reviewer: ```suggestion
> [!IMPORTANT]
```

> > Avoid giving your functions an `id` or `name` that is a built-in Excel function in another language as this could conflict with localized functions.

Reviewer: Since this will conflict, let's remove the "could".
```suggestion
> Avoid giving your functions an `id` or `name` that is a built-in Excel function in another language as this conflicts with localized functions.
```

> The following JSON shows how to define a function with the `id` property "MULTIPLY." The `name` and `description` property of the function is localized for German. Each parameter `name` and `description` is also localized for German.

Reviewer: I know technically the period comes after the quotation mark, but since we're referring to a string literal here, let's be explicit.
```suggestion
The following JSON shows how to define a function with the `id` property "MULTIPLY". The `name` and `description` property of the function is localized for German. Each parameter `name` and `description` is also localized for German.
```

```JSON
{
    "id": "MULTIPLY",
    "name": "SUMME",
    "description": "Summe zwei Zahlen",
    "helpUrl": "http://www.contoso.com",
    "result": {
        "type": "number",
        "dimensionality": "scalar"
    },
    "parameters": [
        {
            "name": "eins",
            "description": "Erste Nummer",
            "dimensionality": "scalar"
        },
        {
            "name": "zwei",
            "description": "Zweite Nummer",
            "dimensionality": "scalar"
        },
    ],
}
```

Compare the previous JSON with the following JSON for English.

```JSON
{
    "id": "MULTIPLY",
    "name": "Multiply",
    "description": "Multiplies two numbers",
    "helpUrl": "http://www.contoso.com",
    "result": {
        "type": "number",
        "dimensionality": "scalar"
    },
    "parameters": [
        {
            "name": "one",
            "description": "first number",
            "dimensionality": "scalar"
        },
        {
            "name": "two",
            "description": "second number",
            "dimensionality": "scalar"
        },
    ],
}
```

### Localize your add-in

> After creating a JSON file for each language, update your XML manifest file with an override value for each locale that specifies the URL of each JSON metadata file. The following manifest XML shows a default `en-us` locale with an override JSON file URL for `de-de` (Germany). The **functions-de.json** file contains the localized German function names and ids.

Reviewer: ```suggestion
After creating a JSON file for each language, update your XML manifest file with an override value for each locale that specifies the URL of each JSON metadata file. The following manifest XML shows a default `en-us` locale with an override JSON file URL for `de-de` (Germany). The **functions-de.json** file contains the localized German function names and IDs.
```
